x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java

```diff
@@ -105,7 +108,7 @@ void start() {
             params,
             autodetectExecutorService,
             (reason) -> {
-                setTaskToFailed(reason, ActionListener.wrap(t -> {}, f -> {}));
+                setTaskToFailed(reason, ActionListener.wrap(t -> {}, task::markAsFailed));
```

Comment:
This is not the source of the bug shown by the test results, but it is definitely one that could occur in a strange scenario. Its best to kill the task all together if setting it to failed failed.

Reply:
`failed` is a state that exists on the persistent task entry in the persistent tasks metadata.  If the whole task is set to failed then that whole entry in the persistent tasks metadata is removed.  So there's no point updating it to have a status of `failed`, then removing it altogether.  We should do one or the other.

For datafeeds there is no `failed` state - if they fail their persistent tasks metadata entry is removed which implies a `stopped` state.

For anomaly detection jobs the idea behind the `failed` state was that it was a way to keep the persistent tasks metadata entry while recording the failure.  This was done because then the `reason` in the persistent tasks metadata entry can record more information about _why_ the task failed and which node it failed on.  This is really useful stuff to know when the likely cause of failure is something that went wrong with a native process.  Not so useful if everything's in Java land (which is the datafeed case).

In the equivalent place for anomaly detection jobs we set the persistent task state to `failed` but keep the persistent task: https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java#L685

Reply:
> So there's no point updating it to have a status of failed, then removing it altogether. We should do one or the other.

This particular case is when we failed to set the persistent task state to failed. Consequently, it would still indicate its running even though it isn't. That is why I closed it (removing from cluster state) when it failed to set failed.

I can look into refactoring setting to failed vs. not existing at all, but I am not sure we want to do that along with this particular bug fix.

Reply:
Sorry, I misread the line.  Yes, what you have is fine, and `markAsFailed` will log the exception that meant we couldn't set the task state to failed.